File: bfinter.rb
program = gets
program = program.chomp
memorysize = 65536
arr = Array.new(memorysize, 0)
pointer = 0
bra = 0
i = 0
while i < program.length do
  if program[i] == '+' then
    if 0 <= pointer and pointer < memorysize then
      arr[pointer] += 1
    else
      puts "Out of Array Bounds At:" + i.to_s
    end
    i += 1
  elsif program[i] == '-' then
    if 0 <= pointer and pointer < memorysize then
      arr[pointer] -= 1
    else
      puts "Out of Array Bounds At:" + i.to_s
    end
    i += 1
  elsif program[i] == '<' then
    pointer -= 1
    i += 1
  elsif program[i] == '>' then
    pointer += 1
    i += 1
  elsif program[i] == ',' then
    inp = gets.to_i
    arr[pointer] = inp
    i += 1
  elsif program[i] == '.' then
    print arr[pointer].chr
    i += 1
  elsif program[i] == '[' then
    if arr[pointer] == 0 then
      bra = 1
      while bra > 0
        i += 1
        if i < program.length and program[i] == '[' then
          bra += 1
        elsif i < program.length and program[i] == ']' then
          bra -= 1
        elsif i >= program.length then
          puts "EOF Reached"
          return
        end
      end
    else
      i += 1
    end
  elsif program[i] == ']' then
    if arr[pointer] != 0 then
      bra = 1
      while bra > 0
        i -= 1
        if i >= 0 and program[i] == '[' then
          bra -= 1
        elsif i >= 0 and program[i] == ']' then
          bra += 1
        elsif i < 0 then
          puts "Out Of Program."
          return
        end
      end
    else
      i += 1
    end
  else
    i += 1
  end
end

File: bfinter.kt
import java.util.*

fun main(args: Array<String>) {
    val sc = Scanner(System.`in`)
    val program = sc.nextLine()
    val memorySize = 65536  //64KB
    val arr = IntArray(memorySize) { 0 }
    var pointer = 0
    var bra: Int = 0
    var i = 0
    while (i < program.length) {
        if (program[i] == '+') {
            if (pointer in 0 until memorySize) {
                arr[pointer]++
            } else {
                println("Out Of Array Bounds At: $i")
                return
            }
            i++
        } else if (program[i] == '-') {
            if (pointer in 0 until memorySize) {
                arr[pointer]--
            } else {
                println("Out Of Array BoundsAt: $i")
                return
            }
            i++
        } else if (program[i] == '<') {
            pointer--
            i++
        } else if (program[i] == '>') {
            pointer++
            i++
        } else if (program[i] == '[') {
            if (arr[pointer] == 0) {
                bra = 1
                while (bra > 0) {
                    i++
                    if (i < memorySize && program[i] == '[') {
                        bra++
                    } else if (i < memorySize && program[i] == ']') {
                        bra--
                    } else if (i >= memorySize) {
                        println("EOF Reached.")
                        return
                    } else {
                    }
                }
            } else {
                i++
            }
        } else if (program[i] == ']') {
            if (arr[pointer] != 0) {
                bra = 1
                while (bra > 0) {
                    i--
                    if (i >= 0 && program[i] == '[') {
                        bra--
                    } else if (i >= 0 && program[i] == ']') {
                        bra++
                    } else if (i < 0) {
                        println("Out Of Program.")
                        return
                    } else {
                    }
                }
            } else {
                i++
            }
        } else if (program[i] == ',') {
            var inp = sc.nextInt()
            arr[pointer] = inp
            i++
        } else if (program[i] == '.') {
            print(arr[pointer].toChar())
            i++
        } else {
            //Other characters.
        }
    }
}
